File: src/SemanticTypeForgers/forger_tests.zig
const std = @import("std");
const forger = @import("forger.zig");
const PrimitiveValue = forger.PrimitiveValue;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const behavior = forger.AtomicBehavior(i64).init("test");

    if (behavior.getPrimitiveType(5) != 5) return error.TestFailed;

    const r1 = try behavior.convertToPrimitive(allocator, PrimitiveValue{ .String = " 123 " });
    if (r1.Int != 123) return error.TestFailed;

    const r2 = try behavior.convertToPrimitive(allocator, PrimitiveValue{ .String = "true" });
    if (r2.Bool != true) return error.TestFailed;

    const r3 = try behavior.convertToPrimitive(allocator, PrimitiveValue{ .String = "{}" });
    if (r3.Int != 0) return error.TestFailed;

    var map = std.StringHashMap(PrimitiveValue).init(allocator);
    defer map.deinit();

    try map.put("productPrice", PrimitiveValue{ .String = "10.5" });
    try map.put("deliveryPrice", PrimitiveValue{ .Float = 2.5 });

    const r4 = try behavior.processValueMap(allocator, map);
    if (r4.Float != 13.0) return error.TestFailed;

    const stdout = std.io.getStdOut().writer();
    try stdout.print("Zig tests passed\n", .{});
}

File: src/SemanticTypeForgers/forger_tests.rs
mod forger;
use forger::{AtomicBehavior, PrimitiveValue};
use std::collections::HashMap;

fn main() {
    let behavior = AtomicBehavior::<i64>::new("test");

    assert_eq!(behavior.get_primitive_type(5), 5);

    let r1 = behavior.convert_to_primitive(&PrimitiveValue::String(" 123 ".to_string()));
    assert_eq!(r1, PrimitiveValue::Int(123));

    let r2 = behavior.convert_to_primitive(&PrimitiveValue::String("true".to_string()));
    assert_eq!(r2, PrimitiveValue::Bool(true));

    let r3 = behavior.convert_to_primitive(&PrimitiveValue::String("{}".to_string()));
    assert_eq!(r3, PrimitiveValue::Int(0));

    let mut map = HashMap::new();
    map.insert("productPrice", PrimitiveValue::String("10.5".to_string()));
    map.insert("deliveryPrice", PrimitiveValue::Float(2.5));

    let r4 = behavior.process_value_map(&map);
    assert_eq!(r4, PrimitiveValue::Float(13.0));

    println!("Rust tests passed");
}
